src/services/offer.js

```diff
@@ -192,29 +196,108 @@ class OfferService {
+     *
      * @param {*} options
      * value: Text to use in full-text-search
-     * offset: Point to start looking (and limiting)
+     * queryToken: Id of the last fetched offer
      * limit: How many offers to show
      * jobType: Array of jobTypes allowed
      */
-    get({ value = "", offset = 0, limit = OfferService.MAX_OFFERS_PER_QUERY, showHidden = false, showAdminReason = false, ...filters }) {
+    async get({ value = "", queryToken = null, limit = OfferService.MAX_OFFERS_PER_QUERY,
+        showHidden = false, showAdminReason = false, ...filters }) {
 
-        const offers = (value ? Offer.find(
-            { "$and": [this._buildFilterQuery(filters), { "$text": { "$search": value } }] }, { score: { "$meta": "textScore" } }
-        ) : Offer.find(this._buildFilterQuery(filters))).current();
+        let offers, queryValue = value, queryFilters = filters;
 
-        if (!showHidden) offers.withoutHidden();
+        if (queryToken) {
+            const {
+                id: lastOfferId,
+                score: lastOfferScore,
+                ...searchInfo
+            } = this.decodeQueryToken(queryToken);
+
+            [queryValue, queryFilters] = [searchInfo.value, searchInfo.filters];
+
+            offers = this._buildSearchContinuationQuery(lastOfferId, lastOfferScore, queryValue,
+                showHidden, showAdminReason, queryFilters);
+        } else {
+            offers = this._buildInitialSearchQuery(queryValue, showHidden, showAdminReason, queryFilters);
+        }
 
-        const offersQuery = offers
-            .sort(value ? { score: { "$meta": "textScore" } } : undefined)
-            .skip(offset)
+        const results = await offers
+            .sort(queryValue ? { score: { "$meta": "textScore" }, _id: 1 } : { _id: 1 })
             .limit(limit)
         ;
 
-        return showAdminReason ? offersQuery : offersQuery.select("-adminReason");
+        if (results.length > 0) {
+            const lastOffer = results[results.length - 1];
+            return {
+                results,
+                queryToken: this.encodeQueryToken(
+                    lastOffer._id,
+                    lastOffer.score || lastOffer._doc?.score,
+                    queryValue, queryFilters
+                ),
+            };
+        } else {
+            return { results };
+        }
+    }
+
+    /**
+     * Builds an initial search query. Cannot be used when loading more offers.
+     * Otherwise, use _buildSearchContinuationQuery().
+     */
+    _buildInitialSearchQuery(value, showHidden, showAdminReason, filters) {
+        const offers = (value ? Offer.find({ "$and": [
+            this._buildFilterQuery(filters),
+            { "$text": { "$search": value } }
+        ] }, { score: { "$meta": "textScore" } }
+
+        ) : Offer.find(this._buildFilterQuery(filters)));
+
+        offers.current();
+        if (!showHidden) offers.withoutHidden();
+        if (!showAdminReason) offers.select("-adminReason");
```

Comment:
Are these filters being applied?

Shouldn't the .current/.withoutHidden/.select be applied to the returned object? I'm not sure the query is being affected by all those filters (depends on whether the different functions mutate the offers object (query) in-place or not)

Reply:
They are being applied. The behaviour is also tested.
By the way, do you think those 3 repeated lines are enough to extract them to another function? It seemed overkill but I'm not sure

Reply:
@imnotteixeira re-review please (the button still isn't working :/)